### ISSUE TITLE
Dependency graph resolution error with micro versions

### ISSUE THREAD

**Issue by maletor** [bundler/bundler#5031]
I was running `bundle update paperclip` with `gem 'rails', '~> 4.2.7.1'` in my Gemfile. However, bundler has a bad code path for recognizing 4.2.7.1 as greater than or equal to 4.2.0.

By changing the rails dependency to `gem 'rails', '~> 4.2.7'`, I was able to work around the issue.

```
Bundler could not find compatible versions for gem "activesupport":
  In Gemfile:
    paperclip (~> 5.1.0) was resolved to 5.1.0, which depends on
      activemodel (>= 4.2.0) was resolved to 5.0.0.1, which depends on
        activesupport (= 5.0.0.1)

    friendly_id (~> 5.1.0) was resolved to 5.1.0, which depends on
      activerecord (>= 4.0.0) was resolved to 4.2.7.1, which depends on
        activesupport (= 4.2.7.1)

    authlogic (~> 3.4.6) was resolved to 3.4.6, which depends on
      activesupport (>= 3.2)

    paperclip (~> 5.1.0) was resolved to 5.1.0, which depends on
      cocaine (~> 0.5.5) was resolved to 0.5.8, which depends on
        climate_control (< 1.0, >= 0.0.3) was resolved to 0.0.3, which depends on
          activesupport (>= 3.0)

    paperclip (~> 5.1.0) was resolved to 5.1.0, which depends on
      activesupport (>= 4.2.0)
```

**Comment by chrismo** [comment]
can you paste in your `bundle env`?

**Comment by maletor** [comment]
What about building one with just rails and paperclip? That should be enough to highlight the issue. If it's unique to my bundler env we can cross that bridge when we get there, but I'd rather not make it public.

**Comment by segiddins** [comment]
In that case, please reduce this to a Gemfile you can share. Thanks!

**Comment by chrismo** [comment]
If you can also include the other bits of `bundle env` - which will show which version of bundler and other environment settings, that will be helpful.

**Comment by maletor** [comment]
```
Environment

    Bundler   1.13.1
    Rubygems  2.5.1
    Ruby      2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin16]
    Git       2.10.0

Bundler settings

    jobs
      Set for the current user (/Users/maletor/.bundle/config): "7"
    path
      Set for your local app (/Users/maletor/src/bundler-5031/.bundle/config): ".bundle/gems"
    disable_shared_gems
      Set for your local app (/Users/maletor/src/bundler-5031/.bundle/config): "true"

Gemfile

    source 'https://rubygems.org'
    gem 'rails', '~> 4.2.7.1'
    gem 'paperclip', '~> 5.1.0'

Gemfile.lock

    GEM
      remote: https://rubygems.org/
      specs:
        actionmailer (4.2.7.1)
          actionpack (= 4.2.7.1)
          actionview (= 4.2.7.1)
          activejob (= 4.2.7.1)
          mail (~> 2.5, >= 2.5.4)
          rails-dom-testing (~> 1.0, >= 1.0.5)
        actionpack (4.2.7.1)
          actionview (= 4.2.7.1)
          activesupport (= 4.2.7.1)
          rack (~> 1.6)
          rack-test (~> 0.6.2)
          rails-dom-testing (~> 1.0, >= 1.0.5)
          rails-html-sanitizer (~> 1.0, >= 1.0.2)
        actionview (4.2.7.1)
          activesupport (= 4.2.7.1)
          builder (~> 3.1)
          erubis (~> 2.7.0)
          rails-dom-testing (~> 1.0, >= 1.0.5)
          rails-html-sanitizer (~> 1.0, >= 1.0.2)
        activejob (4.2.7.1)
          activesupport (= 4.2.7.1)
          globalid (>= 0.3.0)
        activemodel (4.2.7.1)
          activesupport (= 4.2.7.1)
          builder (~> 3.1)
        activerecord (4.2.7.1)
          activemodel (= 4.2.7.1)
          activesupport (= 4.2.7.1)
          arel (~> 6.0)
        activesupport (4.2.7.1)
          i18n (~> 0.7)
          json (~> 1.7, >= 1.7.7)
          minitest (~> 5.1)
          thread_safe (~> 0.3, >= 0.3.4)
          tzinfo (~> 1.1)
        arel (6.0.3)
        builder (3.2.2)
        climate_control (0.0.3)
          activesupport (>= 3.0)
        cocaine (0.5.8)
          climate_control (>= 0.0.3, < 1.0)
        concurrent-ruby (1.0.2)
        erubis (2.7.0)
        globalid (0.3.7)
          activesupport (>= 4.1.0)
        i18n (0.7.0)
        json (1.8.3)
        loofah (2.0.3)
          nokogiri (>= 1.5.9)
        mail (2.6.4)
          mime-types (>= 1.16, < 4)
        mime-types (3.1)
          mime-types-data (~> 3.2015)
        mime-types-data (3.2016.0521)
        mimemagic (0.3.2)
        mini_portile2 (2.1.0)
        minitest (5.9.1)
        nokogiri (1.6.8)
          mini_portile2 (~> 2.1.0)
          pkg-config (~> 1.1.7)
        paperclip (5.1.0)
          activemodel (>= 4.2.0)
          activesupport (>= 4.2.0)
          cocaine (~> 0.5.5)
          mime-types
          mimemagic (~> 0.3.0)
        pkg-config (1.1.7)
        rack (1.6.4)
        rack-test (0.6.3)
          rack (>= 1.0)
        rails (4.2.7.1)
          actionmailer (= 4.2.7.1)
          actionpack (= 4.2.7.1)
          actionview (= 4.2.7.1)
          activejob (= 4.2.7.1)
          activemodel (= 4.2.7.1)
          activerecord (= 4.2.7.1)
          activesupport (= 4.2.7.1)
          bundler (>= 1.3.0, < 2.0)
          railties (= 4.2.7.1)
          sprockets-rails
        rails-deprecated_sanitizer (1.0.3)
          activesupport (>= 4.2.0.alpha)
        rails-dom-testing (1.0.7)
          activesupport (>= 4.2.0.beta, < 5.0)
          nokogiri (~> 1.6.0)
          rails-deprecated_sanitizer (>= 1.0.1)
        rails-html-sanitizer (1.0.3)
          loofah (~> 2.0)
        railties (4.2.7.1)
          actionpack (= 4.2.7.1)
          activesupport (= 4.2.7.1)
          rake (>= 0.8.7)
          thor (>= 0.18.1, < 2.0)
        rake (11.3.0)
        sprockets (3.7.0)
          concurrent-ruby (~> 1.0)
          rack (> 1, < 3)
        sprockets-rails (3.2.0)
          actionpack (>= 4.0)
          activesupport (>= 4.0)
          sprockets (>= 3.0.0)
        thor (0.19.1)
        thread_safe (0.3.5)
        tzinfo (1.2.2)
          thread_safe (~> 0.1)
    
    PLATFORMS
      ruby
    
    DEPENDENCIES
      paperclip (~> 5.1.0)
      rails (~> 4.2.7.1)
    
    BUNDLED WITH
       1.13.1
```

```
~/s/bundler-5031 ❯❯❯ bundle update paperclip
Fetching gem metadata from https://rubygems.org/..........
Fetching version metadata from https://rubygems.org/..
Fetching dependency metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "activesupport":
  In Gemfile:
    paperclip (~> 5.1.0) was resolved to 5.1.0, which depends on
      activemodel (>= 4.2.0) was resolved to 5.0.0.1, which depends on
        activesupport (= 5.0.0.1)

    paperclip (~> 5.1.0) was resolved to 5.1.0, which depends on
      cocaine (~> 0.5.5) was resolved to 0.5.8, which depends on
        climate_control (< 1.0, >= 0.0.3) was resolved to 0.0.3, which depends on
          activesupport (>= 3.0)

    paperclip (~> 5.1.0) was resolved to 5.1.0, which depends on
      activesupport (>= 4.2.0)

    rails (~> 4.2.7.1) was resolved to 4.2.7.1, which depends on
      activesupport (= 4.2.7.1)
```

**Comment by segiddins** [comment]
I just confirmed that https://github.com/bundler/bundler/pull/5101 will fix this
